### PR TITLE
Sequential panner stabilization

### PR DESCRIPTION
**Pull request by teresaqhoang**
### Motivation and Context

This PR sets SequentialPlanner as the default for Chat Copilot and introduces some changes to help stabilize the experience, including:
- Giving the user the option to skip missing functions errors from the planner.
- Allow the user to retry on invalid plan errors (i.e., invalid JSON, plans returned in non-XML or just in natural language)
- Explicitly highlight points of failures in proposed plan to prompt user to fix before approving / denying

#### webapi
Changes on behavior of below flags:
- AllowRetriesOnInvalidPlan: 
     - If true, ExternalInformationSkill will return plan creation if LLM returns an invalid plan.  Fixed invalid plan retry logic when PlanningException.ErrorCodes.InvalidPlan is thrown. 
- SkipOnMissingFunctionsError: Whether to allow missing functions in the plan on creation then sanitize output. Functions are considered missing if they're not available in the planner's kernel's context.
     - If set to true, the plan will be created with missing functions as no-op steps that are filtered from the final proposed plan.
           - When plan returns, sanitize plan of functions not available in Planner's kernel. Merge context variables into plan parameters. Add flag `$???`. to indicate unknown variables in plan preview.
     - If this is set to false, the plan creation will fail if any functions are missing.

If plan contains function not available in planner's kernel's context, removed function is logged to console
image
image

#### webapp
- On view plan:
     -  Add validation for plan step inputs when edits are allowed (if user approval is required). 
     - Display an error message when inputs contain interpolated variables or fields marked `$???`. 
     - Fix PlanStepInput component to include validation for plan step inputs when edits are allowed (if user approval is required) and to update form value on submit. 
     - Fix PlanViewer component to include updated plan in message property and to update context variables on plan action. 
     - Fix an issue where updating a chat message's content was not properly updating the message in the UI. 
     
    Invalid Inputs
image

After valid edits
image

sequentialPlannerStabilization

Sequential Plan
image

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [Contribution Guidelines](https://github.com/microsoft/copilot-chat/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/copilot-chat/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
~~- [ ] All unit tests pass, and I have added new tests where possible~~
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: webapi/CopilotChat/Options/PlannerOptions.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) Microsoft. All rights reserved.
+﻿/// Copyright (c) Microsoft. All rights reserved.
```
[the remark this answers is not in the document]
Comment (teresaqhoang) [reply webapi/CopilotChat/Options/PlannerOptions.cs:1]: No, will change it, but do you know what the guidance is for using // vs ///? The block comments below use ///

[the remark this answers is not in the document]
Comment (TaoChenOSU) [reply webapi/CopilotChat/Options/PlannerOptions.cs:1]: I believe comments that procede with /// are documentation comments: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments. I guess comments that start with // will not be included in the auto-generated docs.

[the remark this answers is not in the document]
Comment (teresaqhoang) [reply webapi/CopilotChat/Options/PlannerOptions.cs]: Awesome, thanks!!

File: webapi/CopilotChat/Skills/ChatSkills/CopilotChatPlanner.cs [hunk 2]
```diff
@@ -43,21 +55,98 @@ public CopilotChatPlanner(IKernel plannerKernel, PlannerOptions? plannerOptions)
...
+        foreach (var step in plan.Steps)
         {
-            return new SequentialPlanner(this.Kernel).CreatePlanAsync(goal);
+            // Check if function exists in planner's kernel
+            if (this.Kernel.Skills.TryGetFunction(step.SkillName, step.Name, out var function))
+            {
+                availableOutputs.AddRange(step.Outputs);
+
+                // Regex to match variable names
+                Regex variableRegEx = new(@"\$((\w+[_-]*)+)", RegexOptions.Singleline);
```
[the remark this answers is not in the document]
Comment (teresaqhoang) [reply webapi/CopilotChat/Skills/ChatSkills/CopilotChatPlanner.cs]: Added as a const with description and examples

File: webapi/CopilotChat/Skills/ChatSkills/ExternalInformationSkill.cs [hunk 3]
```diff
@@ -119,21 +120,21 @@ public class ExternalInformationSkill
             // Create a plan and set it in context for approval.
             var contextString = string.Join("\n", context.Variables.Where(v => v.Key != "userIntent").Select(v => $"{v.Key}: {v.Value}"));
             Plan? plan = null;
-            int maxRetries = 1;
+            int retriesAvail = 3;
```
[the remark this answers is not in the document]
Comment (teresaqhoang) [reply webapi/CopilotChat/Skills/ChatSkills/ExternalInformationSkill.cs]: Added a MissingFunctionError options subsection

File: webapi/CopilotChat/Options/PlannerOptions.cs [hunk 4]
```diff
@@ -9,6 +9,26 @@ namespace SemanticKernel.Service.CopilotChat.Options;
...
+        /// Flag to indicate if skips are allowed on MissingFunction error
+        /// If set to true, the plan will be created with missing functions as no-op steps that are filtered from the final proposed plan.
+        /// If this is set to false, the plan creation will fail if any functions are missing.
+        /// </summary>
+        public bool AllowRetries { get; set; } = true;
+
+        /// <summary>
+        /// Max retries allowed on MissingFunctionsError.
+        /// </summary>
+        /// [Range(1, 5)]
```
Comment (TaoChenOSU) [inline webapi/CopilotChat/Options/PlannerOptions.cs:28]: comment?

**Review by TaoChenOSU — APPROVED** [review]

**Pull request merged**